clang/lib/CodeGen/CodeGenTypes.cpp

```diff
@@ -118,6 +124,37 @@ llvm::Type *CodeGenTypes::ConvertTypeForMem(QualType T, bool ForBitField) {
   return R;
 }
 
+bool CodeGenTypes::LLVMTypeLayoutMatchesAST(QualType ASTTy,
+                                            llvm::Type *LLVMTy) {
+  CharUnits ASTSize = Context.getTypeSizeInChars(ASTTy);
+  CharUnits LLVMSize =
+      CharUnits::fromQuantity(getDataLayout().getTypeAllocSize(LLVMTy));
+  return ASTSize == LLVMSize;
+}
+
+llvm::Type *CodeGenTypes::convertTypeForLoadStore(QualType T,
+                                                  llvm::Type *LLVMTy) {
+  if (!LLVMTy)
+    LLVMTy = ConvertType(T);
+
+  if (!T->isBitIntType() && LLVMTy->isIntegerTy(1))
+    return llvm::IntegerType::get(getLLVMContext(),
+                                  (unsigned)Context.getTypeSize(T));
+
+  if (T->isBitIntType()) {
+    llvm::Type *R = ConvertType(T);
+    if (!LLVMTypeLayoutMatchesAST(T, R))
+      return llvm::Type::getIntNTy(
+          getLLVMContext(), Context.getTypeSizeInChars(T).getQuantity() * 8);
```

Comment:
`return llvm::Type::getIntNTy(
          getLLVMContext(), Context.getTypeSizeInChars(T).getQuantity() * 8);`
Could you plz explain the logic behind using this implementation ? do you support all types > 128 and also types < 128 ? or it's specific to the types > 128 only ?

Reply:
We intend to support all types. In general, _BitInt(N) is lowered to iN type in LLVM IR. However it is not possible for certain sizes and some targets now due to different alignment defined for iN and _BitInt(N). This now happens for _BitInt(>=129) for x86_64 targets. `LLVMTypeLayoutMatchesAST` checks for this mismatch. If there is mismatch, say NBytes = sizeof(_BitInt(N)), we use [i8 x NBytes] as in-memory type. It is better to load and store value using iNBytes type instead of an array type since it is better for optimization and guarantees that all bits of the loaded/stores value is valid. For more info, please refer to https://github.com/llvm/llvm-project/pull/91364#issuecomment-2099351151 as well.

Reply:
I don't understand why the choice of type is conditional. No matter of the "layout" matches or not, the stores have to produce valid in-memory representation (according to the target ABI), which in the general case means writing **all the** bits of the in-memory representation.

Reply:
The choice is conditional to keep changes less invasive.

I wonder if that requires to have different in-memory type for all _BitInt types?
But in case when the "layout" matches we could use `iNBytes ` for both memory type and load-store type.
Though that makes me wonder what will be the whole point of small `_BitInt` then...

Reply:
cf. https://github.com/llvm/llvm-project/pull/91364#issuecomment-2099384663

I'm fairly certain using load/store type of `iBITS` is the correct thing to do, unconditionally.
Not quite sure about the choice between `iBITS` and `[BYTES x i8]`, if we're not talking about a load/stores how the array type could possibly be less efficient, so we don't default unconditionally to it?

Reply:
We could literally use `[N x i8]` as the allocation type of every single type and save ourselves a lot of grief.  It's nicer to generate more semantic types when we can, though; it produces IR that's easier to both read and test.

Reply:
FWIW, using `[N x i8]` for all types is exactly what Rust does nowadays.

Reply:
I have a concern that this way we will always need to use a different memory type for _BitInt (perhaps always `[N x i8]`). I'm not sure it is right for all targets?

Reply:
I would like to use LLVM's `iN` type when it actually has the right data layout.  If we decide to give up and use a Rust-like opaque approach, we should do it comprehensively and not just for `_BitInt`.

Reply:
Getting back to this...

> I wonder making it unconditional that requires to have different in-memory type for all _BitInt types?

The in-memory representation is a whole number of bytes, so in general the number of in-memory bits will be different from the declared number on a `_BitInt` type.

If you have `_BitInt(K)` and a load/store type `iN`, then loading/storing `iK` won't necessarily write to the same bytes as loading/storing `iN` would. That means we won't be writing to all the bytes of the in-memory representation which might be incorrect (e.g. for AArch32 ABI which requires padding bits to be sign- or zero- extended). It can also be inefficient, because LLVM requires memory accesses to be the minimum number of bytes sufficient to hold an `iK` - e.g. `i17` will end up as a load store of `i16` followed by a load store of `i8`, whereas the in-memory representation is 4 bytes  and can be accesses as a single `i32`.